src/py/flwr/server/superlink/fleet/grpc_rere/server_interceptor.py

```diff
@@ -172,3 +162,64 @@ def _generic_method_handler(
             request_deserializer=method_handler.request_deserializer,
             response_serializer=method_handler.response_serializer,
         )
+
+    def _verify_node_id(
+        self,
+        node_id: Optional[int],
+        request: Union[
+            DeleteNodeRequest,
+            PullTaskInsRequest,
+            PushTaskResRequest,
+            GetRunRequest,
+            PingRequest,
+        ],
+    ) -> bool:
+        if node_id is None:
+            return False
+        if isinstance(request, PushTaskResRequest):
+            if len(request.task_res_list) == 0:
+                return False
+            return request.task_res_list[0].task.consumer.node_id == node_id
+        if isinstance(request, GetRunRequest):
+            return node_id in self.state.get_nodes(request.run_id)
+        return request.node.node_id == node_id
+
+    def _verify_hmac(
+        self, public_key: ec.EllipticCurvePublicKey, request: Request, hmac_value: bytes
+    ) -> bool:
+        shared_secret = generate_shared_key(self.server_private_key, public_key)
+        return verify_hmac(shared_secret, request.SerializeToString(True), hmac_value)
+
+    def _create_authenticated_node(
+        self,
+        public_key_bytes: bytes,
+        request: CreateNodeRequest,
+        context: grpc.ServicerContext,
+        method_handler: grpc.RpcMethodHandler,
+    ) -> CreateNodeResponse:
+        context.send_initial_metadata(
+            (
+                (
+                    _PUBLIC_KEY_HEADER,
+                    self.encoded_server_public_key,
+                ),
+            )
+        )
+        try:
+            node_id = self.state.get_node_id(public_key_bytes)
+        except KeyError:
+            node_id = None
+
+        if node_id is not None:
+            with self._lock:
+                # Handle RPC here instead of passing to default method_handler
+                self.state.restore_node(node_id, request.ping_interval)
+                return CreateNodeResponse(node=Node(node_id=node_id, anonymous=False))
```

Comment:
we need to do the `self.state.get_node_id(public_key_bytes)` check again (within the scope of the lock)